pydicer/analyse/data.py

```diff
@@ -0,0 +1,261 @@
+        self.dataset_name = dataset_name
+        self.dataset_directory = self.data_directory.joinpath(dataset_name)
+
+    def get_all_computed_radiomics_for_dataset(self):
+        """Return a DataFrame of radiomics computed for this dataset
+
+        Returns:
+            pd.DataFrame: The DataFrame of all radiomics computed for dataset
+        """
+
+        dfs = []
+        for radiomics_file in self.dataset_directory.glob("**/structures/*.csv"):
+            dfs.append(pd.read_csv(radiomics_file, index_col=0))
+
+        df = pd.concat(dfs)
+        df.sort_values(["Patient", "Contour"], inplace=True)
+
+        return df
+
+    def compute_radiomics(
+        self,
+        patient=None,
+        radiomics=None,
+        settings=None,
+        structure_meta_data=None,
+        image_meta_data=None,
+        resample_to_image=False,
+    ):
+        """
+        Visualise the data in the working directory. PNG files are generates providing a
+        snapshot of the various data objects.
+
+        Args:
+            patient (list|str, optional): A patient ID (or list of patient IDs) to visualise.
+            Defaults to None.
+            radiomics (dict, optional): A dictionary of the pyradiomics to compute. Format should
+                have the radiomic class name as the key and a list of feature names in the value.
+                See https://pyradiomics.readthedocs.io/en/latest/features.html for more
+                information.
+                Defaults to all first order features.
+            settings (dict, optional): Settings to pass to pyradiomics. Defaults to
+                PYRAD_DEFAULT_SETTINGS.
+            structure_meta_data (list, optional): A list of DICOM tags which will be extracted from
+                the structure DICOM headers and included in the resulting table of radiomics.
+                Defaults to None.
+            image_meta_data ([type], optional): A list of DICOM tags which will be extracted from
+                the image DICOM headers and included in the resulting table of radiomics.
+                Defaults to None.
+            resample_to_image (bool, optional): Define if mask should be resampled to image. If not
+                the image will be resampled to mask. Defaults to False.
+
+        Raises:
+            ValueError: Raised if patient is not None, a list of strings or a string.
+        """
+        if isinstance(patient, list):
+            if not all(isinstance(x, str) for x in patient):
+                raise ValueError("All patient IDs must be of type 'str'")
+        elif patient is None:
+            patient = [p.name for p in self.dataset_directory.glob("*") if p.is_dir()]
+        else:
+
+            if not isinstance(patient, str) and patient is not None:
+                raise ValueError(
+                    "Patient ID must be list or str. None is valid to process all patients"
+                )
+            patient = [patient]
+
+        if radiomics is None:
```

Comment:
Could we set these default values to be in the function definition, rather than having ```None``` as the default then checking if these are ```None```? So like:
```python
def compute_radiomics(
        self,
        patient=None,
        radiomics=DEFAULT_RADIOMICS,
        settings=PYRAD_DEFAULT_SETTINGS,
        structure_meta_data=[],
        image_meta_data=[],
        resample_to_image=False,
    ):
```

Reply:
I did have that, but pylint gives a warning that that isn't secure for some reason... Not actually sure why but I thought I'd just make pylint happy for now :D Will look into what exactly the reason is :)

Reply:
Ah ok I see why its bad practice. So if you use a dict or list (both mutable) as a default parameter, then this same default object uses the same object across multiple calls to the function. So if it changes during the function call then it will have the changed object in the next function all instead of starting with a fresh default object.

This example should demonstrate it:
```python
def func_bad(k, v, para={}):
    para[k] = v
    return para
    
print(func_bad('a', 1)) #returns {'a': 1}
print(func_bad('b', 2)) #returns {'a': 1, 'b': 2}
print(func_bad('c', 3)) #returns {'a': 1, 'b': 2, 'c': 3}
    
    
def func_good(k, v, para=None):
    if para is None:
        para = dict()
    para[k] = v
    return para
    
print(func_good('a', 1)) #returns {'a': 1}
print(func_good('b', 2)) #returns {'b': 2}
print(func_good('c', 3)) #returns {'c': 3}
```

Reply:
Ah interesting never knew that, great to know it now :D